File: backend/app/models/game.py
from datetime import datetime
from typing import Literal, Optional

from pydantic import BaseModel, Field


class LastMove(BaseModel):
    from_square: str = Field(alias="from")
    to: str

    model_config = {
        "populate_by_name": True,
    }


class GameSettings(BaseModel):
    skill_level: int
    move_time: int
    depth: int


class GameSnapshot(BaseModel):
    fen: str
    moves: list[str] = Field(default_factory=list)
    player_color: Literal["white", "black"]
    evaluation: float = 0
    evaluation_history: list[dict] = Field(default_factory=list)
    last_move: Optional[dict] = None
    game_status: Literal[
        "playing",
        "check",
        "checkmate",
        "draw",
        "stalemate",
    ] = "playing"
    winner: Optional[Literal["white", "black"]] = None
    settings: GameSettings


class ActiveGameResponse(GameSnapshot):
    id: str
    status: Literal["active", "completed"]
    updated_at: datetime


class PlayerStatsResponse(BaseModel):
    games_played: int
    active_games: int
    completed_games: int
    wins: int
    losses: int
    draws: int
    ai_requests: int
    total_tokens: int
    level: int

File: frontend/types/game.ts
import { EvaluationPoint } from '@/types/evaluation';
import { EngineSettings } from '@/types/engine';

export interface LastMove {
  from: string;
  to: string;
}

export interface GameSnapshot {
  fen: string;
  moves: string[];
  player_color: 'white' | 'black';
  evaluation: number;
  evaluation_history: EvaluationPoint[];
  last_move: LastMove | null;
  game_status:
    | 'playing'
    | 'check'
    | 'checkmate'
    | 'draw'
    | 'stalemate';
  winner: 'white' | 'black' | null;
  settings: EngineSettings;
}

export interface ActiveGame extends GameSnapshot {
  id: string;
  status: 'active' | 'completed';
  updated_at: string;
}

export interface PlayerStats {
  games_played: number;
  active_games: number;
  completed_games: number;
  wins: number;
  losses: number;
  draws: number;
  ai_requests: number;
  total_tokens: number;
  level: number;
}
